### PR TITLE
[Profiler] Add request profiling to performance metrics

### PR DESCRIPTION
**Pull request by rpkamp**
| Q                | A
| ---------------- | -----
| Bug fix          | no
| New feature      | yes
| BC breaks        | no
| Deprecations     | no
| Tests pass       | yes
| Fixed tickets    | #192
| License          | MIT

### CODE REVIEW

File: src/Middleware/ProfileMiddleware.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace EightPoints\Bundle\GuzzleBundle\Middleware;
+
+use Symfony\Component\Stopwatch\Stopwatch;
+
+class ProfileMiddleware
+{
+    /**
+     * @var Stopwatch
```
Comment (gregurco) [inline src/Middleware/ProfileMiddleware.php]: please use full class name in comments:
```
@var \Symfony\Component\Stopwatch\Stopwatch
```

File: src/Middleware/ProfileMiddleware.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+     */
+    private $stopwatch;
+
+    /**
+     * ProfileMiddleware constructor.
```
Comment (gregurco) [inline src/Middleware/ProfileMiddleware.php]: `ProfileMiddleware constructor.` - please remove this comment line

File: src/DependencyInjection/EightPointsGuzzleExtension.php [hunk 3]
```diff
@@ -251,6 +257,31 @@ protected function defineLogMiddleware(ContainerBuilder $container, Definition $
...
+     *
+     * @param \Symfony\Component\DependencyInjection\ContainerBuilder $container
+     * @param \Symfony\Component\DependencyInjection\Definition $handler
+     * @param string $clientName
+     *
+     * @return void
+     */
+    protected function defineProfileMiddleware(ContainerBuilder $container, Definition $handler, string $clientName)
+    {
+        if (!class_exists('Symfony\Component\Stopwatch\Stopwatch')) {
```
[the remark this answers is not in the document]
Comment (rpkamp) [reply src/DependencyInjection/EightPointsGuzzleExtension.php]: It is actually in the require-dev of symfony/framework-bundle, which is in the require of this package, so it's already in require-dev implicitly. This `if` is just to get a clearer message when people try to run this prod for whatever reason.

File: src/DependencyInjection/EightPointsGuzzleExtension.php [hunk 4]
```diff
@@ -110,7 +112,7 @@ public function load(array $configs, ContainerBuilder $container)
      *
      * @return \Symfony\Component\DependencyInjection\Definition
      */
-    protected function createHandler(ContainerBuilder $container, string $clientName, array $options, bool $logging) : Definition
+    protected function createHandler(ContainerBuilder $container, string $clientName, array $options, bool $logging, bool $profiling) : Definition
```
[the remark this answers is not in the document]
Comment (rpkamp) [reply src/DependencyInjection/EightPointsGuzzleExtension.php]: done

File: src/DependencyInjection/EightPointsGuzzleExtension.php [hunk 5]
```diff
@@ -251,6 +257,31 @@ protected function defineLogMiddleware(ContainerBuilder $container, Definition $
...
+     *
+     * @return void
+     */
+    protected function defineProfileMiddleware(ContainerBuilder $container, Definition $handler, string $clientName)
+    {
+        if (!class_exists('Symfony\Component\Stopwatch\Stopwatch')) {
+            throw new RuntimeException('Eightpoint GuzzleBundle profiling requires "symfony/stopwatch" package');
+        }
+
+        $logMiddlewareDefinitionName = sprintf('eight_points_guzzle.middleware.profile.%s', $clientName);
```
[the remark this answers is not in the document]
Comment (rpkamp) [reply src/DependencyInjection/EightPointsGuzzleExtension.php]: done

File: src/Middleware/ProfileMiddleware.php (hunk 1 above)
Comment (rpkamp) [reply src/Middleware/ProfileMiddleware.php]: done

File: src/Middleware/ProfileMiddleware.php (hunk 2 above)
Comment (rpkamp) [reply src/Middleware/ProfileMiddleware.php]: done

File: src/DependencyInjection/EightPointsGuzzleExtension.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (rpkamp) [reply src/DependencyInjection/EightPointsGuzzleExtension.php]: I tried that, but for some reason `$container->hasDefinition('debug.stopwatch')` always returns `false`. Maybe something to do with the order in which things are loaded?

Checking if the class exists seemed like the next best thing.

**Pull request merged**